File: 28.搜索二维矩阵/Solution.py
class Solution:
    """
    @param matrix, a list of lists of integers
    @param target, an integer
    @return a boolean, indicate whether matrix contains target
    """
    def searchMatrix(self, matrix, target):
        # write your code here
        m = len(matrix)
        if m == 0:
            return False
        n = len(matrix[0])
        if n == 0:
            return False
        i, j = 0, m-1
        while i < j:
            mid = (i + j)/2
            if mid == i:
                mid = j
            if matrix[mid][0] == target:
                return True
            elif matrix[mid][0] < target:
                i = mid
            elif matrix[mid][0] > target:
                j = mid - 1
        k = i
        i, j = 0, n-1
        while i < j:
            mid = (i + j)/2
            if matrix[k][mid] == target:
                return True
            elif matrix[k][mid] < target:
                i = mid + 1
            elif matrix[k][mid] > target:
                j = mid - 1
        if matrix[k][i] == target:
            return True
        return False

File: 28.搜索二维矩阵/Solution.cpp
class Solution {
public:
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        // write your code here
        int m = matrix.size();
        if(m == 0)
            return false;        
            
        int n = matrix[0].size();

        int l = 0, r = m*n-1;
        int mid = (r + l)/2;
        int row = 0, col = 0;
        while(l < r){
            row = mid/n;
            col = mid%n;
            if(matrix[row][col] == target)
                break;
            else if(matrix[row][col] > target)
                r = mid - 1;
            else
                l = mid + 1;
                
            mid = (l + r)/2;
        }
        
        row = mid/n;
        col = mid%n;
        if(matrix[row][col] != target || row < 0 || col < 0)
            return false;
        else
            return true;
        
    }
};
